### DISCUSSION TITLE
Final push to release LightKrylov v1.0.0

### DISCUSSION THREAD

**Discussion by loiseaujc** [nekStab/LightKrylov#45] [General]
## Where are we ?

`LightKrylov` is almost ready to reach v1.0.0 ! So far, it provides the following functionalities:
- `AbstractVector` : module providing the abstract definition of a vector.
  + `abstract_vector` is the base type all other pieces of software need to extend to be compatible with `LightKrylov`. Only a handful of type-bound procedures are declared as `deferred` and need to be implemented by the user for their particular definition of a vector.
- `LinearOperator` : module providing the abstract definition of a linear operator.
  + `abstract_linop` is the base type from which all others are being extended. To define a derived type from it, one needs to provide the implement for `matvec` (i.e. `y = A @ x`) and `rmatvec` (i.e. `y = A.T @ x`).
  + `abstract_spd_linop` extends `abstract_linop` for symmetric positive (semi-) definite operators.
  + `identity_linop` is, as the name suggest, the identity operator.
  + `scaled_linop` is a derived type to construct a new operator `Y = aX` with `a` a real number.
  + `axpby_linop` is a derived type to construct a new operator `Z = aX + bY` with `a` and `b` two real numbers.
- `BaseKrylov` : module providing a large collection of Krylov-like factorizations for abstract linear operators. It includes :
  + `arnoldi_factorization` - the go-to factorization for computing the leading eigenvalues of a linear operator.
  + `lanczos_tridiagonlization` - specialization of the Arnoldi factorization when `A` is symmetric.
  + `lanczos_bidiagonalization` - iterative solver to compute the leading singular triplets of an abstract linear operator.
  + `nonsymmetric_lanczos_tridiagonalization` - Reduction of an arbitrary operator (`n x n`) to a nonsymmetric tridiagonal matrix by simultaneously constructing two bi-orthogonal Krylov bases for the image and the domain of the operator. Note that it is numerically unstable and is only included for the sake of completeness.
  + `two_sided_arnoldi_factorization` - Similar to `nonsymmetric_lanczos_tridiagonalization` but construct two orthonormal bases for the image and the domain of the operator. It is far more stable numerically and should be preferred.
  + `qr_factorization` - Modified Gram-Schmidt procedure to orthonormalize an arbitrary Krylov basis. @Simkern : I didn't check, but you reimplemented it to make all the computations in-place, right?
- `IterativeSolvers` : module built on top of `BaseKrylov` providing several Krylov-based iterative solvers.
  + `eigs` - Compute the leading eigenvalues and eigenvectors of an arbitrary real linear operator.
  + `eighs` - Compute the leading eigenvalues and eigenvectors of a symmetric real linear operator.
  + `svds` - Compute the leading singular triplets of an arbitrary linear operator.
  + `gmres` - Solve `Ax = b` using the *generalized minimum residual` method. `A` is an arbitrary linear operator.
  + `cg` - Solve `Ax = b` using the *conjugate gradient* method. `A` is a symmetric positive definite linear operator.

There are a couple of other methods implemented such as the `two_sided_eigs` or `rational_krylov_factorization` but these are still experimental for the moment.

---

## What is needed to reach v1.0.0 ?

As it is, `LightKrylov` already covers quite a large set of applications. We have also been able to couple it with a Ginzburg-Landau solver as well as with `Nek5000` (MPI-based CFD solver). Expect maybe for a bit of cleaning, I think we have almost everything we'll need for v1.0.0. Here is a short list of what I think is still needed on the side to have a proper release:
- `example` : Implement a minimal set of examples illustrating how to use `LightKrylov` in a somewhat realistic application. I plan to port some of the Ginzburg-Landau code base here to illustrate the computation of the leading eigenpairs in a time-stepper approach. If you have other ideas, they are most welcomed.
- `documentation` : the in-code documentation is already pretty good although we do need to write `docstrings` for all functions (see `gmres` for an example). It's not the most interesting part of the job but it shouldn't take long. We also need to setup a proper online documentation website. @ricardofrantz : some ideas on which framework we could use for that ?
- `JOSS` : I think writing a *Journal of Open Source Software* on `LightKrylov` would be a good idea. We need to check exactly what is required by JOSS to be able to submit paper + repo for review.

@ricardofrantz and @Simkern : Do you see anything that is missing and/or should be implemented for v1.0.0 ? Timing-wise, I can take a large part of next week to address most of the things listed here (except for actually writing the JOSS manuscript, we're ain't quite there yet). Once `LightKrylov` has reached v1.0.0, we can then focus our attention on rewriting `neklab` and continuing the development of `LightROM`. As we work on the other packages, we'll eventually see if other things need to be implemented here or not.

What do you think?

**Comment by Simkern** [comment]
Hej!
`LightKrylov` looks really good already!

Some comments from my side:
`qr_factorization`: Yes, the code does a simple no-frills double GS. I have indeed updated it to be in-place. I propagated this change to the tests as well. As we had discussed, the code could, in due time, be improved with pivoting and breakdown checks if we see that the current implementation is insufficient. I will expand the `docstring` in the style of `gmres`.
`documentation`: I agree that more documentation would be helpful. In particular, I think a diagram showing the dependencies of the functions as well as the types throughout the toolbox would be very helpful. We could use something like Sphinx [https://www.sphinx-doc.org/en/master/].

Features for `LightKrylov`/`LightROM`: A number of applications of `LightKrylov` will involve more operations of the `abstract_vector` types, including projections onto the Krylov bases. For now, `LightKrylov` has the necessary definitions (in particular `get_vec`). I have been adding some methods into `LightROM` (e.g. the marix product between Krylov bases). The organisational question would be whether all the basic methods related to the types defined in `LightKrylov` should be gathered here (for clarity) or included in the repo where they are directly used (making each package individually as light as possible).

**Reply by loiseaujc** [reply]
`documentation` : I'll start working on it right away. I'll add the docstrings to all the major subroutines over the next couple of days. Properly documenting the repo also implies:
- Updating the README with a better description of what `LightKrylov` is aiming at as well as adding some sections about references, guidelines for contributions, etc. I'll look at some nice repos later today to see how they did.
- Online documentation : I've never used `sphinx` before although I've seen it deployed quite a few times. @Simkern and @ricardofrantz : any of you already has some experience with it?
- Testing with more compilers : At the moment, the Github Actions only tests that the code base compiles with `gcc`/`gfortran`. It'd be nice if we could also test with intel's compilers as they are available on many HPC facilities. I don't have it installed on my laptop though and I'm not sure how to include it in the github actions/fpm framework. It is not the most urgent thing to do though. @ricardofrantz : If you have time, could you look into this? I seem to recall that you have `ifort` on your laptop.
- Examples : at the moment, I've added an example computing the leading eigenvalues of the linearized ginzburg-landau equation (see #48). Maybe an example using `mpi` to showcase that it works just the same with a parallel time-stepper would also be a good idea. To keep it simple, the example could be as simple as computing the leading eigenpairs of a 2D Laplace operator on the unit square (using time-stepping again).

Regarding `LightKrylov` vs `LightROM` : I think having everything acting on low-level objects such as `abstract_vector` or `abstract_linop` should be included in `LightKrylov` eventhough, at the moment, only `LightROM` would make use of these utilities. Like so, we have a clear scope separation :
- `LightKrylov` : provides all the low-level objects (`abstract_vector`, `abstract_linop`) and associated procedures (`dot`, `get_tmat`, etc) as well as the base Krylov methods (`arnoldi`, `eigs`, etc).
- `LightROM` : built on top of `LightKrylov`, it provides all of the functionalities needed to study LTI systems.

The primary reason is that, if we develop another packages that needs something like `get_mat` but not all the LTI stuff, we only need to add `LightKrylov` as deps and not both `LightKrylov` and `LightROM`.

**Reply by Simkern** [reply]
`documentation`:

- `sphinx`: Unfortunately, I haven't used `sphinx` myself, I've just seen it in many places and a colleague from KTH recommended it. 
- Examples: I merged the GL code now. I think the 2D Laplacian is a good addition, but I think also the 1D Laplacian is good because it is much smaller for testing purposes. I'll add it to `LightROM` for now to test DLRA.

`LightKrylov` vs. `LightROM`: I agree entirely, I would do it the same way. For now the routines are in LightROM in my dev branch but we can move them over. The product routines are going to facilitate resolving #6 as well as extending #35 to block QR updates.

**Reply by loiseaujc** [reply]
I'll look into how `sphinx` works next week then and see how to incorporate into the workflow. In the mean time, I'll look into nice README from other repos to see how we can update ours. This is probably the only I'll be able to do today as Maud's teacher is sick and I have to look after kid n°1.
